File: core/python/pubplot/recipes/paired_compare.py
"""Recipe: paired before/after comparison. Mirrors the R recipe: each subject
(--id) is measured under two conditions (--x), and a line connects the subject's
two points. The paired t-test or the Wilcoxon signed-rank test is chosen from the
normality of the within-subject differences."""
import numpy as np
import pandas as pd
from scipy import stats as ss

from ..bundle import slugify
from ..theme import apply_pub_style, new_fig, palette
from ..util import cap_first, p_stars, shapiro_normal


def recipe_paired_compare(df, spec):
    x, y, idc = spec["data"]["x"], spec["data"]["y"], spec["data"].get("id")
    if not x or not y:
        raise ValueError("paired_compare needs --x (condition) and --y (value)")
    if not idc:
        raise ValueError("paired_compare needs --id (the subject column)")
    for c in (idc, x, y):
        if c not in df.columns:
            raise ValueError(f"column '{c}' not found in data")

    d = df[[idc, x, y]].copy()
    d[y] = pd.to_numeric(d[y], errors="coerce")
    d = d.dropna()
    d[x] = d[x].astype(str)
    lv = list(pd.unique(d[x]))  # preserve appearance order (pre before post)
    if len(lv) != 2:
        raise ValueError(f"paired_compare needs exactly 2 conditions in '{x}'; found {len(lv)}")

    a = d[d[x] == lv[0]][[idc, y]].drop_duplicates(subset=idc).rename(columns={y: "v1"})
    b = d[d[x] == lv[1]][[idc, y]].drop_duplicates(subset=idc).rename(columns={y: "v2"})
    m = a.merge(b, on=idc)
    dropped = d[idc].nunique() - len(m)
    if len(m) < 2:
        raise ValueError("paired_compare needs at least 2 complete pairs")
    steps = [f"Kept {len(m)} complete pair(s) across the two conditions."]
    if dropped:
        steps.append(f"Dropped {dropped} subject(s) missing one condition.")

    diffs = (m["v2"] - m["v1"]).to_numpy()
    pdiff, dnorm = shapiro_normal(diffs)
    method = (spec["test"].get("method") or "auto").lower()
    if method in ("auto", ""):
        family = "t" if dnorm else "wilcoxon"
    elif method in ("t", "t_test", "paired_t", "student_t"):
        family = "t"
    elif method in ("wilcoxon", "wilcox", "signed_rank", "mwu"):
        family = "wilcoxon"
    else:
        raise ValueError(f"unknown paired test '{method}' (use t or wilcoxon)")

    if family == "t":
        res = ss.ttest_rel(m["v2"], m["v1"])
        stat = float(res.statistic); pval = float(res.pvalue); dof = len(m) - 1
        eff = float(np.mean(diffs) / np.std(diffs, ddof=1)); eff_name = "Cohen's dz"
        label = "paired t-test"
    else:
        nz = diffs[diffs != 0]
        res = ss.wilcoxon(m["v2"], m["v1"])
        stat = float(res.statistic); pval = float(res.pvalue); dof = np.nan
        rk = ss.rankdata(np.abs(nz))
        eff = float((rk[nz > 0].sum() - rk[nz < 0].sum()) / rk.sum()) if len(nz) else np.nan
        eff_name = "rank-biserial r"; label = "Wilcoxon signed-rank test"

    ap = spec["appearance"]
    pal = palette(2, ap.get("palette"))
    w, h = 3.8, 4.2
    fig, ax = new_fig(w, h)
    for _, r in m.iterrows():
        ax.plot([0, 1], [r["v1"], r["v2"]], color="grey", lw=0.6, alpha=0.7, zorder=1)
    ax.scatter([0] * len(m), m["v1"], color=pal[0], s=45, zorder=2, alpha=0.9)
    ax.scatter([1] * len(m), m["v2"], color=pal[1], s=45, zorder=2, alpha=0.9)
    ax.set_xticks([0, 1]); ax.set_xticklabels(lv); ax.set_xlim(-0.4, 1.4)

    allv = np.r_[m["v1"].to_numpy(), m["v2"].to_numpy()]
    yr = float(np.nanmax(allv) - np.nanmin(allv)) or 1.0
    ytop = float(np.nanmax(allv)) + 0.08 * yr
    blabel = (ap.get("bracket") or {}).get("label", "p.signif")
    brk = f"p = {pval:.2g}" if blabel == "p.format" else p_stars(pval)
    ax.plot([0, 0, 1, 1], [ytop, ytop + 0.02 * yr, ytop + 0.02 * yr, ytop], color="black", lw=1.0)
    ax.text(0.5, ytop + 0.03 * yr, brk, ha="center", va="bottom", fontsize=12)

    ax.set_xlabel(ap.get("x_label") or x)
    ax.set_ylabel(ap.get("y_label") or y)
    if ap.get("title"):
        ax.set_title(ap["title"], fontweight="bold")
    apply_pub_style(ax, ap.get("theme"))
    fig.tight_layout()

    stats_df = pd.DataFrame([{
        "outcome": y, "condition1": lv[0], "condition2": lv[1], "n_pairs": len(m),
        "test": label, "statistic": stat, "df": dof, "p_value": pval,
        "mean_diff": float(np.mean(diffs)), "effect_size_name": eff_name,
        "effect_size": eff, "shapiro_p_diff": pdiff,
    }])
    test_meta = {
        "name": label, "statistic": stat, "p_value": pval, "n_pairs": len(m),
        "mean_difference": float(np.mean(diffs)),
        "effect_size": {"name": eff_name, "value": eff},
        "assumptions": {"normality_of_differences": {"test": "shapiro-wilk", "p": pdiff}},
    }
    methods = _methods(label, lv, len(m), eff_name, eff, spec)
    return {
        "fig": fig, "stats": stats_df, "test_meta": test_meta, "methods": methods,
        "df_used": m, "clean_steps": steps, "label": slugify(y),
        "width": w, "height": h,
        "qc": {"quantity": "within-subject differences", "panels": {"differences": diffs}},
        "build_script": lambda in_name, fig_stub: _script(spec, family, lv, in_name, fig_stub),
    }


def _methods(label, lv, n, eff_name, eff, spec):
    from importlib.metadata import version as v
    import platform
    y_lab = spec["appearance"].get("y_label") or spec["data"]["y"]
    return (
        f"{cap_first(y_lab)} was compared between {lv[0]} and {lv[1]} within each subject with a "
        f"{label} (n = {n} pairs). The test was chosen from the Shapiro-Wilk normality of the "
        f"within-subject differences. Effect size is reported as {eff_name} = {eff:.2f}. "
        "Significance was set at P < 0.05. "
        f"Analyses were performed in Python {platform.python_version()} with scipy {v('scipy')}; "
        f"the figure was produced with matplotlib {v('matplotlib')}."
    )


def _script(spec, family, lv, in_name, fig_stub):
    x, y, idc = spec["data"]["x"], spec["data"]["y"], spec["data"]["id"]
    test_line = ('res = ss.ttest_rel(m["v2"], m["v1"])' if family == "t"
                 else 'res = ss.wilcoxon(m["v2"], m["v1"])')
    return f'''#!/usr/bin/env python3
# Standalone reproduction. Run inside the pinned pubplot container.
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np, pandas as pd
from scipy import stats as ss

raw = pd.read_csv("{in_name}")
raw["{y}"] = pd.to_numeric(raw["{y}"], errors="coerce")
raw = raw.dropna(subset=["{idc}", "{x}", "{y}"])
raw["{x}"] = raw["{x}"].astype(str)
lv = {lv}
a = raw[raw["{x}"] == lv[0]][["{idc}", "{y}"]].drop_duplicates("{idc}").rename(columns={{"{y}": "v1"}})
b = raw[raw["{x}"] == lv[1]][["{idc}", "{y}"]].drop_duplicates("{idc}").rename(columns={{"{y}": "v2"}})
m = a.merge(b, on="{idc}")
{test_line}
print(res)
fig, ax = plt.subplots(figsize=(3.8, 4.2))
for _, r in m.iterrows(): ax.plot([0, 1], [r["v1"], r["v2"]], color="grey", lw=0.6, alpha=0.7)
ax.scatter([0]*len(m), m["v1"], s=45); ax.scatter([1]*len(m), m["v2"], s=45)
ax.set_xticks([0, 1]); ax.set_xticklabels(lv); ax.set_xlabel("{x}"); ax.set_ylabel("{y}")
for s in ("top", "right"): ax.spines[s].set_visible(False)
fig.tight_layout(); fig.savefig("{fig_stub}.pdf")
print("Reproduced figure.")
'''

File: core/r/R/paired_compare.R
# Recipe: paired before/after comparison. Each subject (--id) is measured under
# two conditions (--x); a line connects the subject's two points. The paired
# t-test or the Wilcoxon signed-rank test is chosen from the normality of the
# within-subject differences, unless forced with --test.

recipe_paired_compare <- function(df, spec) {
  x <- spec$data$x; y <- spec$data$y; idc <- spec$data$id
  if (is.null(x) || is.null(y)) stop("paired_compare needs --x (condition) and --y (value)")
  if (is.null(idc) || !nzchar(idc)) stop("paired_compare needs --id (the subject column)")
  for (cn in c(idc, x, y)) if (!cn %in% names(df)) stop(sprintf("column '%s' not found in data", cn))

  df <- df[, c(idc, x, y)]
  df[[y]] <- suppressWarnings(as.numeric(df[[y]]))
  n_rows0 <- nrow(df)
  df <- df[stats::complete.cases(df), , drop = FALSE]
  # Preserve the order the conditions appear in (so pre stays left of post),
  # rather than sorting alphabetically.
  df[[x]] <- factor(df[[x]], levels = unique(as.character(df[[x]])))
  lv <- levels(df[[x]])
  if (length(lv) != 2) {
    stop(sprintf("paired_compare needs exactly 2 conditions in '%s'; found %d (%s)",
                 x, length(lv), paste(lv, collapse = ", ")))
  }

  a <- df[df[[x]] == lv[1], c(idc, y)]; names(a) <- c("id", "v1")
  b <- df[df[[x]] == lv[2], c(idc, y)]; names(b) <- c("id", "v2")
  a <- a[!duplicated(a$id), ]; b <- b[!duplicated(b$id), ]
  m <- merge(a, b, by = "id")
  n_ids <- length(unique(df[[idc]]))
  dropped <- n_ids - nrow(m)
  if (nrow(m) < 2) stop("paired_compare needs at least 2 complete pairs")
  clean_steps <- c(
    sprintf("Kept %d complete pair(s) across the two conditions.", nrow(m)),
    if (dropped > 0) sprintf("Dropped %d subject(s) missing one condition.", dropped) else NULL)

  diffs <- m$v2 - m$v1
  norm  <- shapiro_safe(diffs)
  method <- tolower(spec$test$method %||% "auto")
  family <- if (method %in% c("auto", "")) {
    if (isTRUE(norm$normal)) "t" else "wilcoxon"
  } else if (method %in% c("t", "t_test", "paired_t", "student_t")) "t"
    else if (method %in% c("wilcoxon", "wilcox", "signed_rank", "mwu")) "wilcoxon"
    else stop(sprintf("unknown paired test '%s' (use t or wilcoxon)", method))

  if (family == "t") {
    tt <- stats::t.test(m$v2, m$v1, paired = TRUE)
    stat <- unname(tt$statistic); pval <- tt$p.value; dfree <- unname(tt$parameter)
    eff <- mean(diffs) / stats::sd(diffs); eff_name <- "Cohen's dz"; label <- "paired t-test"
  } else {
    # exact = NULL lets R take the exact route when no tie and no zero remain,
    # which is the rule the Python engine repeats.
    wt <- suppressWarnings(stats::wilcox.test(m$v2, m$v1, paired = TRUE))
    stat <- unname(wt$statistic); pval <- wt$p.value; dfree <- NA_real_
    eff <- rank_biserial_paired(diffs)
    eff_name <- "rank-biserial r"; label <- "Wilcoxon signed-rank test"
  }

  plong <- rbind(
    data.frame(id = m$id, cond = lv[1], v = m$v1, stringsAsFactors = FALSE),
    data.frame(id = m$id, cond = lv[2], v = m$v2, stringsAsFactors = FALSE))
  plong$cond <- factor(plong$cond, levels = lv)

  pal <- spec$appearance$palette %||% pubready_palette[seq_len(2)]
  star <- if (is.na(pval)) "ns" else if (pval < 0.001) "***" else
          if (pval < 0.01) "**" else if (pval < 0.05) "*" else "ns"
  blabel <- spec$appearance$bracket$label %||% "p.signif"
  brk <- if (blabel == "p.format") paste0("p = ", formatC(pval, format = "g", digits = 2)) else star
  ypos <- max(plong$v, na.rm = TRUE) + 0.06 * diff(range(plong$v, na.rm = TRUE))

  p <- ggplot2::ggplot(plong, ggplot2::aes(x = cond, y = v)) +
    ggplot2::geom_line(ggplot2::aes(group = id), colour = "grey70",
                       linewidth = 0.4, alpha = 0.8) +
    ggplot2::geom_point(ggplot2::aes(colour = cond), size = 2.4, alpha = 0.9) +
    ggplot2::scale_colour_manual(values = pal) +
    ggpubr::geom_bracket(xmin = lv[1], xmax = lv[2], y.position = ypos,
                         label = brk, tip.length = 0.02) +
    ggplot2::labs(x = spec$appearance$x_label %||% x,
                  y = spec$appearance$y_label %||% y,
                  title = spec$appearance$title) +
    pub_theme(spec$appearance$theme)

  stats_df <- data.frame(
    outcome = y, condition1 = lv[1], condition2 = lv[2], n_pairs = nrow(m),
    test = label, statistic = stat, df = dfree, p_value = pval,
    mean_diff = mean(diffs), effect_size_name = eff_name, effect_size = eff,
    shapiro_p_diff = norm$p, stringsAsFactors = FALSE)

  test_meta <- list(
    name = label, statistic = stat, p_value = pval,
    n_pairs = nrow(m), mean_difference = mean(diffs),
    effect_size = list(name = eff_name, value = eff),
    assumptions = list(normality_of_differences = list(test = "shapiro-wilk", p = norm$p)))

  resolved <- list(family = family, label = label)
  methods  <- .paired_methods(label, lv, nrow(m), pval, eff_name, eff, spec)
  build_script <- function(in_name, fig_stub) .paired_emit_script(spec, family, lv, in_name, fig_stub)

  list(plot = p, stats = stats_df, test_meta = test_meta, resolved = resolved,
       methods = methods, build_script = build_script,
       label = slugify(y), df_used = m, clean_steps = clean_steps,
       qc = list(quantity = "within-subject differences",
                 panels = list(differences = diffs)),
       width = 3.8, height = 4.2)
}

.paired_methods <- function(label, lv, n, pval, eff_name, eff, spec) {
  v    <- pkg_versions(c("ggplot2", "ggpubr"))
  rver <- paste(R.version$major, R.version$minor, sep = ".")
  y_lab <- spec$appearance$y_label %||% spec$data$y
  paste0(
    sprintf("%s was compared between %s and %s within each subject with a %s (n = %d pairs). ",
            cap_first(y_lab), lv[1], lv[2], label, n),
    "The test was chosen from the Shapiro-Wilk normality of the within-subject differences. ",
    sprintf("Effect size is reported as %s = %s. Significance was set at P < 0.05. ",
            eff_name, formatC(eff, format = "f", digits = 2)),
    sprintf("Drawn in R %s with ggplot2 %s and ggpubr %s.", rver, v$ggplot2, v$ggpubr))
}

.paired_emit_script <- function(spec, family, lv, in_name, fig_stub) {
  x <- spec$data$x; y <- spec$data$y; idc <- spec$data$id
  test_line <- if (family == "t") {
    "res <- t.test(m$v2, m$v1, paired = TRUE)"
  } else "res <- suppressWarnings(wilcox.test(m$v2, m$v1, paired = TRUE))"
  c(
    "#!/usr/bin/env Rscript",
    "# Standalone reproduction. Run inside the pinned pubready container.",
    "suppressPackageStartupMessages({ library(ggplot2); library(ggpubr) })",
    "",
    sprintf('raw <- read.csv("%s", check.names = FALSE)', in_name),
    sprintf('raw[["%s"]] <- as.numeric(raw[["%s"]]); raw <- raw[complete.cases(raw[, c("%s","%s","%s")]), ]',
            y, y, idc, x, y),
    sprintf('a <- raw[raw[["%s"]] == "%s", c("%s","%s")]; names(a) <- c("id","v1")', x, lv[1], idc, y),
    sprintf('b <- raw[raw[["%s"]] == "%s", c("%s","%s")]; names(b) <- c("id","v2")', x, lv[2], idc, y),
    "m <- merge(a[!duplicated(a$id), ], b[!duplicated(b$id), ], by = \"id\")",
    test_line,
    "print(res)",
    sprintf('pl <- rbind(data.frame(id=m$id, cond="%s", v=m$v1), data.frame(id=m$id, cond="%s", v=m$v2))',
            lv[1], lv[2]),
    sprintf('pl$cond <- factor(pl$cond, levels = c("%s","%s"))', lv[1], lv[2]),
    "p <- ggplot(pl, aes(cond, v)) + geom_line(aes(group = id), colour = \"grey70\") +",
    "  geom_point(aes(colour = cond), size = 2.4) +",
    sprintf('  labs(x = "%s", y = "%s") + theme_classic() + theme(legend.position = "none")', x, y),
    sprintf('ggsave("%s.pdf", p, width = 3.8, height = 4.2)', fig_stub),
    'cat("Reproduced figure.\\n")'
  )
}
